src/mds/MDSRankQuiesce.cc

```diff
@@ -0,0 +1,592 @@
+            ++it;
+          } else {
+            // just ack.
+            dout(20) << "INACTIVE RESPONDER: reporting '" << it->first << "' as " << it->second.state << dendl;
+            it = quiesce_map.roots.erase(it);
+          }
+          break;
+        default:
+          it = quiesce_map.roots.erase(it);
+          break;
+        }
+      }
+      return true;
+    });
+
+    if (quiesce_agent) {
+      // reset the agent if it's present
+      // because it won't receive any more callbacks
+      quiesce_agent->reset_async();
+    }
+  } else {
+    rebind_agent_callback(quiesce_agent, quiesce_db_manager);
+  }
+
+  quiesce_db_manager->update_membership(membership, inject_request);
+}
+
+bool MDSRank::quiesce_dispatch(const cref_t<Message> &m) {
+  switch(m->get_type()) {
+    case MSG_MDS_QUIESCE_DB_LISTING:
+    {
+      const auto& req = ref_cast<MMDSQuiesceDbListing>(m);
+      if (quiesce_db_manager) {
+        dout(10) << "got " << req->db_listing << " from peer " << req->gid << dendl;
+        int result = quiesce_db_manager->submit_listing_from(req->gid, std::move(req->db_listing));
+        if (result != 0) {
+          dout(3) << "error (" << result << ") submitting " << req->db_listing << " from peer " << req->gid << dendl;
+        }
+      } else {
+        dout(5) << "no db manager to process " << req->db_listing << dendl;
+      }
+      return true;
+    }
+    case MSG_MDS_QUIESCE_DB_ACK:
+    {
+      const auto& req = ref_cast<MMDSQuiesceDbAck>(m);
+      if (quiesce_db_manager) {
+        dout(10) << "got ack " << req->diff_map << " from peer " << req->gid << dendl;
+        int result = quiesce_db_manager->submit_ack_from(req->gid, std::move(req->diff_map));
+        if (result != 0) {
+          dout(3) << "error (" << result << ") submitting an ack from peer " << req->gid << dendl;
+        }
+      } else {
+        dout(5) << "no db manager to process an ack: " << req->diff_map << dendl;
+      }
+      return true;
+    }
+    default:
+      return false;
+  }
+}
+
+void MDSRank::quiesce_agent_setup() {
+  // TODO: replace this with a non-debug implementation
+  //       Potentially, allow the debug interface under some runtime configuration
+
+  ceph_assert(quiesce_db_manager);
+
+  using RequestHandle = QuiesceInterface::RequestHandle;
+  using QuiescingRoot = std::pair<RequestHandle, Context*>;
+  auto quiesce_requests = std::make_shared<std::unordered_map<QuiesceRoot, QuiescingRoot>>();
+
+  QuiesceAgent::ControlInterface ci;
+
+  ci.submit_request = [this, quiesce_requests](QuiesceRoot root, Context* c)
+      -> std::optional<RequestHandle> {
+    auto uri = boost::urls::parse_uri_reference(root);
+    if (!uri) {
+      dout(5) << "error parsing the quiesce root as an URI: " << uri.error() << dendl;
+      c->complete(uri.error());
+      return std::nullopt;
+    } else {
+      dout(20) << "parsed root '" << root <<"' as : " << uri->path() << " " << uri->query() << dendl;
+    }
```

Comment:
```suggestion
    }
    dout(5) << "submit_request: " << uri << dendl;
```

I think is sufficient. Use higher debugging for entry points to major APIs.

Reply:
Ack. I've combined this with another dout I had right before taking the mds lock